src/code.cloudfoundry.org/loggregator/plumbing/envelope_averager.go

```diff
@@ -0,0 +1,81 @@
+package plumbing
+
+import (
+	"sync/atomic"
+	"time"
+)
+
+// EnvelopeAverager implements google.golang.org/grpc/stats.Handler and keeps
+// track of the average envelope size emitted by a grpc client. It should
+// be constructed with NewEnvelopeAverager.
+//
+// Note: Track is on a very "hot" path.
+// Care should be taken while altering any algorithms, and performance
+// must be considered.
+type EnvelopeAverager struct {
+	// storage must be accessed via atomics
+	storage countAndTotal
+}
+
+// NewEnvelopeAverager creates a new EnvelopeAverager.
+func NewEnvelopeAverager() *EnvelopeAverager {
+	return &EnvelopeAverager{}
+}
+
+// Track takes the given envelope size (in bytes) to use in the current
+// average calculation. It can be called by several go-routines.
+func (a *EnvelopeAverager) Track(envelopeSize int) {
+	for {
+		current := atomic.LoadUint64((*uint64)(&a.storage))
+		newValue := countAndTotal(current).encodeDelta(1, uint32(envelopeSize))
+
+		if atomic.CompareAndSwapUint64((*uint64)(&a.storage), current, uint64(newValue)) {
+			return
+		}
+	}
+}
+
+// Start invokes the given callback at the set interval.
+func (a *EnvelopeAverager) Start(interval time.Duration, f func(average float64)) {
+	go func() {
+		var prevCount, prevTotal uint32
+		for range time.Tick(interval) {
+			current := atomic.LoadUint64((*uint64)(&a.storage))
```

Comment:
I don't quite follow why you would have to do a CAS here vs a single `SwapUint64`. `SwapUint64` would cause some CAS operations in `Track` to fail but that isn't a big deal if the interval is at 60 seconds. I am worried that with just 64 bits to work with for both count and total you risk one of them overflowing if you do not reset to 0. With 48 bits dedicated to the total you likely will not overflow total but with only 16 bits for count you can now only count up to 65536 which you will certainly hit very frequently. What is the behavior when that happens? I don't know off the top of my head. I would take the risk of overflow out by doing the swap with 0. Perhaps you can explain to me why you feel that's a bad idea tomorrow.